From fec07cb76fe99c6a6d4123fc50bce90702ceb66b Mon Sep 17 00:00:00 2001
From: Vasily Khoruzhick <anarsoul@gmail.com>
Date: Thu, 24 May 2012 15:02:47 +0300
Subject: [PATCH 3/3] Fix content get_width/height

---
 content/content.c |    6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/content/content.c b/content/content.c
index f3345db..8daab3a 100644
--- a/content/content.c
+++ b/content/content.c
@@ -984,6 +984,9 @@ const char *content__get_status_message(struct content *c)
  */
 int content_get_width(hlcache_handle *h)
 {
+	if (h == NULL)
+		return 0;
+
 	return content__get_width(hlcache_handle_get_content(h));
 }
 
@@ -1003,6 +1006,9 @@ int content__get_width(struct content *c)
  */
 int content_get_height(hlcache_handle *h)
 {
+	if (h == NULL)
+		return 0;
+
 	return content__get_height(hlcache_handle_get_content(h));
 }
 
-- 
1.7.10.2

