Index: netsurf/Makefile.defaults
===================================================================
--- netsurf.orig/Makefile.defaults	2011-08-29 21:04:11.281463001 +1000
+++ netsurf/Makefile.defaults	2011-08-29 21:04:56.681463001 +1000
@@ -82,8 +82,8 @@
 # Valid options: YES, NO
 NETSURF_USE_LIBICONV_PLUG := YES
 
-# Initial CFLAGS. Optimisation level etc. tend to be target specific.
-CFLAGS := 
+# Take CFLAGS from environment
+# CFLAGS := 
 
 # Default installation/execution prefix
 PREFIX ?= /usr/local
