diff -crB imgv-0.3.0-orig/Makefile imgv-0.3.0//Makefile
*** imgv-0.3.0-orig/Makefile	2012-02-14 08:27:48.355688611 -0600
--- imgv-0.3.0//Makefile	2012-02-14 08:31:10.698222827 -0600
***************
*** 1,13 ****
  ####### openwrt-xburst path
  
  BASEPATH      = /home/fcarello/Projects/openwrt-xburst
! TARGETPATH    ?= $(BASEPATH)/staging_dir/target-mipsel_uClibc-0.9.32
  
  ####### Compiler, tools and options
  
  CC            = mipsel-openwrt-linux-gcc
  STRIP         = mipsel-openwrt-linux-strip
! CFLAGS        := $(CFLAGS) -pipe -O2 -fomit-frame-pointer -mips32 -mtune=mips32 -funit-at-a-time -fhonour-copts -msoft-float -Wall -W -D_REENTRANT
  INCPATH       = -I$(TARGETPATH)/usr/include/ -I$(TARGETPATH)/usr/include/SDL -I.
  LIBS          = -L$(TARGETPATH)/usr/lib/ -lSDL -lSDL_image -lSDL_gfx -lSDL_ttf -lfreetype -ldirectfb -ldirect -lfusion -lz -lpthread
  
--- 1,13 ----
  ####### openwrt-xburst path
  
  BASEPATH      = /home/fcarello/Projects/openwrt-xburst
! TARGETPATH    ?= $(BASEPATH)/staging_dir/target-arm_v5te__uClibc-0.9.32
  
  ####### Compiler, tools and options
  
  CC            = mipsel-openwrt-linux-gcc
  STRIP         = mipsel-openwrt-linux-strip
! CFLAGS        := $(CFLAGS)
  INCPATH       = -I$(TARGETPATH)/usr/include/ -I$(TARGETPATH)/usr/include/SDL -I.
  LIBS          = -L$(TARGETPATH)/usr/lib/ -lSDL -lSDL_image -lSDL_gfx -lSDL_ttf -lfreetype -ldirectfb -ldirect -lfusion -lz -lpthread
  
