*** imgv.orig/Makefile	2012-09-26 08:38:31.000000000 -0400
--- imgv-0.3.0/Makefile	2012-09-26 08:34:16.000000000 -0400
***************
*** 1,15 ****
  ####### openwrt-xburst path
  
  BASEPATH      = /home/fcarello/Projects/openwrt-xburst
! TARGETPATH    ?= $(BASEPATH)/staging_dir/target-mipsel_uClibc-0.9.32
  
  ####### Compiler, tools and options
  
  CC            = mipsel-openwrt-linux-gcc
  STRIP         = mipsel-openwrt-linux-strip
! CFLAGS        := $(CFLAGS) -pipe -O2 -fomit-frame-pointer -mips32 -mtune=mips32 -funit-at-a-time -fhonour-copts -msoft-float -Wall -W -D_REENTRANT
  INCPATH       = -I$(TARGETPATH)/usr/include/ -I$(TARGETPATH)/usr/include/SDL -I.
! LIBS          = -L$(TARGETPATH)/usr/lib/ -lSDL -lSDL_image -lSDL_gfx -lSDL_ttf -lfreetype -ldirectfb -ldirect -lfusion -lz -lpthread
  
  SOURCE1       = sdl-imageviewer.c
  TARGET1       = imgv
--- 1,15 ----
  ####### openwrt-xburst path
  
  BASEPATH      = /home/fcarello/Projects/openwrt-xburst
! TARGETPATH    ?= $(BASEPATH)/staging_dir/target-arm_v5te__uClibc-0.9.32
  
  ####### Compiler, tools and options
  
  CC            = mipsel-openwrt-linux-gcc
  STRIP         = mipsel-openwrt-linux-strip
! CFLAGS        := $(CFLAGS)
  INCPATH       = -I$(TARGETPATH)/usr/include/ -I$(TARGETPATH)/usr/include/SDL -I.
! LIBS          = -L$(TARGETPATH)/usr/lib/ -lSDL -lSDL_image -lSDL_gfx -lSDL_ttf -lfreetype -lz -lpthread
  
  SOURCE1       = sdl-imageviewer.c
  TARGET1       = imgv
