Index: fheroes2/src/dist/Makefile
===================================================================
--- fheroes2/src/dist/Makefile	(revision 2861)
+++ fheroes2/src/dist/Makefile	(working copy)
@@ -6,7 +6,10 @@
 PREFIX := /usr/local
 endif
 
+CFLAGS := $(CFLAGS) -DCONFIGURE_FHEROES2_LOCALEDIR=\"$(PREFIX)/share/fheroes2\"
+CFLAGS := $(CFLAGS) -DCONFIGURE_FHEROES2_DATA=\"$(PREFIX)/share/fheroes2\"
 TARGET := fheroes2
+
 LIBENGINE := ../engine/libengine.a
 CFLAGS := $(CFLAGS) -I../engine
 
@@ -15,10 +18,6 @@
 CFLAGS := $(CFLAGS) -I../xmlccwrap
 endif
 
-ifdef PREFIX
-CFLAGS := $(CFLAGS) -I$(PREFIX)/include
-endif
-
 RES	:=
 ICOFILE	:= ../../image/fheroes2_32x32.ico
 
Index: fheroes2/Makefile
===================================================================
--- fheroes2/Makefile	(revision 2861)
+++ fheroes2/Makefile	(working copy)
@@ -23,7 +23,7 @@
 MAKE	:= make
 
 all:
-	$(MAKE) -C src WITH_AI=empty
+	$(MAKE) -C src WITH_AI=simple RELEASE=1
 	@cp src/dist/$(TARGET) .
 
 clean:
