*** ./net_main.c	1999-12-21 17:53:26.000000000 -0600
--- ../chumby_quake/sdlquake-1.0.9/net_main.c	2008-07-19 03:50:28.000000000 -0500
***************
*** 872,880 ****
  	// initialize all the drivers
  	for (net_driverlevel=0 ; net_driverlevel<net_numdrivers ; net_driverlevel++)
  		{
  		controlSocket = net_drivers[net_driverlevel].Init();
! 		if (controlSocket == -1)
  			continue;
  		net_drivers[net_driverlevel].initialized = true;
  		net_drivers[net_driverlevel].controlSock = controlSocket;
  		if (listening)
--- 872,884 ----
  	// initialize all the drivers
  	for (net_driverlevel=0 ; net_driverlevel<net_numdrivers ; net_driverlevel++)
  		{
+         Con_DPrintf("Initializing driver %s...\n", net_drivers[net_driverlevel].name);
  		controlSocket = net_drivers[net_driverlevel].Init();
! 		if (controlSocket == -1) {
!             Con_DPrintf("Failed to init network driver %s\n", net_drivers[net_driverlevel].name);
  			continue;
+         }
+         Con_DPrintf("Successfully initted network driver %s\n", net_drivers[net_driverlevel].name);
  		net_drivers[net_driverlevel].initialized = true;
  		net_drivers[net_driverlevel].controlSock = controlSocket;
  		if (listening)
