*** ./net_dgrm.c	1999-12-21 17:53:26.000000000 -0600
--- ../chumby_quake/sdlquake-1.0.9/net_dgrm.c	2008-07-19 03:53:16.000000000 -0500
***************
*** 776,784 ****
  
  	for (i = 0; i < net_numlandrivers; i++)
  		{
  		csock = net_landrivers[i].Init ();
! 		if (csock == -1)
  			continue;
  		net_landrivers[i].initialized = true;
  		net_landrivers[i].controlSock = csock;
  		}
--- 776,788 ----
  
  	for (i = 0; i < net_numlandrivers; i++)
  		{
+         Con_DPrintf("Initting dgram driver %s...\n", net_landrivers[i].name);
  		csock = net_landrivers[i].Init ();
! 		if (csock == -1) {
!             Con_DPrintf("Failed to init dgram driver %s.\n", net_landrivers[i].name);
  			continue;
+         }
+         Con_DPrintf("Successfully initted dgram driver %s.\n", net_landrivers[i].name);
  		net_landrivers[i].initialized = true;
  		net_landrivers[i].controlSock = csock;
  		}
